drivers/pci/controller/vmd.c

```diff
@@ -767,17 +831,239 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
 	return 0;
 }
 
-static int vmd_enable_domain(struct vmd_dev *vmd, unsigned long features)
+static void vmd_configure_cfgbar(struct vmd_dev *vmd)
+{
+	struct resource *res;
+	u16 ioc_bus_range = 0;
+	u16 pch_bus_range = 0;
+
+	res = &vmd->dev->resource[VMD_CFGBAR];
+	vmd->resources[VMD_RESOURCE_CFGBAR] = (struct resource){
+		.name = "VMD CFGBAR",
+		.start = vmd->busn_start[VMD_ROOTBUS_0],
+		.end = vmd->busn_start[VMD_ROOTBUS_0] +
+		       (resource_size(res) >> 20) - 1,
+		.flags = IORESOURCE_BUS | IORESOURCE_PCI_FIXED,
+	};
+
+	if (vmd_has_pch_rootbus(vmd)) {
+		pci_read_config_word(vmd->dev, PCI_REG_BUSRANGE0,
+				     &ioc_bus_range);
+		pci_read_config_word(vmd->dev, PCI_REG_BUSRANGE1,
+				     &pch_bus_range);
+
+		/*
+		 * Resize CPU IOC CFGBAR range to make space for PCH owned
+		 * devices by adjusting range end with value stored in
+		 * PCI_REG_BUSRANGE0 register.
+		 */
+		vmd->resources[VMD_RESOURCE_CFGBAR].start =
+			VMD_ROOTBUS_RANGE_START(ioc_bus_range);
+		vmd->resources[VMD_RESOURCE_CFGBAR].end =
+			VMD_ROOTBUS_RANGE_END(ioc_bus_range);
+
+		vmd->resources[VMD_RESOURCE_PCH_CFGBAR] = (struct resource){
+			.name = "VMD CFGBAR PCH",
+			.start = VMD_ROOTBUS_RANGE_START(pch_bus_range),
+			.end = VMD_ROOTBUS_RANGE_END(pch_bus_range),
+			.flags = IORESOURCE_BUS | IORESOURCE_PCI_FIXED,
+			.parent = &vmd->resources[VMD_RESOURCE_CFGBAR],
+		};
+	}
+}
+
+/**
+ * vmd_configure_membar - Configure VMD MemBAR register, which points
+ * to MMIO address assigned by the OS or BIOS.
+ * @vmd: the VMD device
+ * @resource_number: resource buffer number to be filled in
+ * @membar_number: type of the MemBAR
+ * @start_offset: 4K aligned offset applied to start of VMD’s MEMBAR MMIO space
+ * @end_offset: 4K aligned offset applied to end of VMD’s MEMBAR MMIO space
+ * @parent: parent resource assigned to resource to be filled in
+ *
+ * Function fills resource buffer inside the VMD structure.
+ */
+static void vmd_configure_membar(struct vmd_dev *vmd,
+				 enum vmd_resource resource_number,
+				 u8 membar_number, resource_size_t start_offset,
```

Comment:
``` suggestion
 				enum vmd_resource membar_number, resource_size_t start_offset,
```

Reply:
membar number  does not have an enum, it is a standalone define.